Implement -I, --md-selector, and --no-prelude in kprove, kprovex, keq, and kbmc

SchmErik opened the pull request:
This change implements command line options from OuterParsingOptions class for several tools other than kompile. This is broken down into the following changes:

Extract `-d` from `OuterParsingOptions`
----------------------------------------------------
This extraction prepares `OuterParsingOptions` class to be included in the
following tools: kprove, kprovex, keq, and kmbc.

KProve, KEq, KBMC: add OuterParsing options
----------------------------------------------------
Adding this object enables the `-I`, `--md-selector`, and `--no-prelude`
options to these tools. By doing so, each tool can configure their own
OuterParsing options rather than getting this information by de-serializing
the kompile object from the cache left by a previous invocation of the
`kompile` command.

pyk tests: remove -I from kompile add -I to kprove
----------------------------------------------------
This test case utilized -I in the kompile command in order to pass the
parameter of this option to kprove. This was done through serializing
the `kompile` object during the invocation of the `kompile` command and
de-serializing the `kompile` object to access the `outerParsingOption`
during the invocation of the `kprove` command. Since the -I option has
been added to `kprove`, drop this command from the `kompile` command and
move it to the `kprove` command.

kprove-markdown test: use --md-selector for testing
-------------------------------------------------------
Add additional lines in a test file to test the new --md-selector
option for kprove. Before this feature was added, the default behavior was for
kprove to use `k` which is the default md-selector. In order to test that the --md-selector
option is functional, text has been added to the test case so that it will
cause a failure on the default md-selector.



Fixes: #1909
Fixes: #1330

Commit:
Extract -d from OuterParsingOptions

This extraction prepares OuterParsingOptions class to be included in the
following tools: kprove, kprovex, keq, and kmbc.

Commit:
KProve, KEq, KBMC: add OuterParsing options

Adding this object enables the `-I`, `--md-selector`, and `--no-prelude`
options to these tools. By doing so, each tool can configure their own
OuterParsing options rather than getting this information by de-serializing
the kompile object from the cache left by a previous invocation of the
`kompile` command.

Commit:
pyk tests: remove -I from kompile add -I to kprove

This test case utilized -I in the kompile command in order to pass the
parameter of this option to kprove. This was done through serializing
the `kompile` object during the invocation of the `kompile` command and
de-serializing the `kompile` object to access the `outerParsingOption`
during the invocation of the `kprove` command. Since the -I option has
been added to `kprove`, drop this command from the `kompile` command and
move it to the `kprove` command.

radumereuta commented on kernel/src/main/java/org/kframework/kompile/Kompile.java:
```diff
@@ -86,20 +87,32 @@
+         */
+        if (outerParsing != null) {
+            kompileOptions.outerParsing = outerParsing;
+        }
```
@dwightguth can you please have a look here?
On my PR you requested changes for something similar.

dwightguth replied:
I would prefer instead of mutating KompileOptions that we simply store the OuterParsingOptions in the Kompile object and reference it directly when we need to read its values. We try to avoid this type of mutability in the code base where possible.

SchmErik replied:
I think the main problem that I see is that kompileOptions.outerParsing is initialized differently depending on the command that we're running. If I can find a way to construct an instance of kompile options with the appropriate outerParsing object, we can eliminate this change along with the constructor change

radumereuta reviewed (commented):
Overall very good. Just one comment where I would like @dwightguth to have a look.
Also, the comments in the first message seem to be a repeat of the commit messages.
I wouldn't include them there. If I need further explanations for each commit, then I can have a look at each one. Save the first message for an overall description of what you are trying to achieve in this PR. In this case I wouldn't even bother since the issue you are trying to fix already describes the problem pretty well.

dwightguth commented on kernel/src/main/java/org/kframework/kompile/Kompile.java:
```diff
@@ -86,20 +87,32 @@
     java.util.Set<KEMException> errors;
 
     public Kompile(KompileOptions kompileOptions, FileUtil files, KExceptionManager kem, boolean cacheParses) {
```
where is this constructor called? I would prefer not to construct an instance with an uninitialized field like this.

SchmErik replied:
this is being called from k-distribution/src/test/scala/org/kframework/backend/kore/KoreTest.scala line 41:
val compiler = new Kompile(options, files, kem, false)

SchmErik replied:
@dwightguth I guess one option is to initialize the outerParsingOption with kompileOptions.OuterParsing but I am not sure if we should be doing that. This constructor is called from a test and the other one is called from kompile so I'm tempted to say that this would be fine...

dwightguth commented on kernel/src/main/java/org/kframework/kompile/Kompile.java:
```diff
@@ -86,20 +87,32 @@
     }
 
     public Kompile(KompileOptions kompileOptions, FileUtil files, KExceptionManager kem, Stopwatch sw, boolean cacheParses) {
+        this(kompileOptions, null, files, kem, sw, cacheParses);
```
same question here

SchmErik replied:
kernel/src/main/java/org/kframework/kompile/KompileFrontEnd.java line 67

dwightguth reviewed (changes requested):
seems mostly reasonable but I had a few specific comments

Commit:
add code review suggestions

radumereuta commented:
The final version should probably contain a test that exercises the `-I` and `--md-selector` options.

SchmErik commented:
> The final version should probably contain a test that exercises the `-I` and `--md-selector` options.

You helped me add the -I tests by helping me fix the pyk tests so I think we all that we have left is a good --md-selector test. Do you know of any good test cases to exercise this?

Commit:
Compiled definition: add outerParsingOptions

Commit:
KompileFrontEnd: use injected OuterParsingOptions

Commit:
kprove-markdown: use --md-selector for testing

This test adds additional lines in a test file to test the new --md-selector
option for kprove. Before this feature was added, the default behavior was for
kprove to use k as the selector. In order to test that the --md-selector
option is functional, text has been added to the test case so that it will
cause a failure on the default md-selector.

dwightguth reviewed (commented):
Looks good except you should probably use KProveOptions.outerParsing when constructing the CompiledDefinition object in kprove.

Commit:
Kprove: use KproveOptions.outerparsing to initialize kompiledDefinition

Commit:
Merge remote-tracking branch 'upstream/master' into kprove-i-2

Commit:
Update branch with master

radumereuta commented on kernel/src/main/java/org/kframework/kompile/KompileOptions.java:
```diff
@@ -38,6 +39,9 @@ GlobalOptions getGlobalOptions_UseOnlyInGuiceProvider() {
 
+    @ParametersDelegate
+    public transient OutputDirectoryOptions outputDirectory = new OutputDirectoryOptions();
+
```
I'm a bit confused. Wasn't OuterParsingOptions supposed to be transient, and OutputDirectoryOptions be serializable?

Commit:
KompileOptions: make OuterparsingOptions transient

radumereuta commented on kernel/src/main/java/org/kframework/kdep/KDepModule.java:
```diff
@@ -36,7 +38,8 @@ GlobalOptions globalOptions(KDepOptions options) {
+    OuterParsingOptions outerParsingOptions(KDepOptions options) { return options.outerParsing; }
+
+    @Provides
+    OutputDirectoryOptions outputDirectoryOptions(KompileOptions options) { return options.outputDirectory; }
```
@dwightguth shouldn't these be `@RequestScoped`?

dwightguth replied:
It honestly doesn't matter. The only difference is that if it's request scoped, this function will be called only once instead of each time the options are requested. But this function is so simple it will be inlined by the jvm, so there's no significant difference in performance because of this.

radumereuta commented on kernel/src/main/java/org/kframework/kompile/KompileModule.java:
```diff
@@ -45,4 +46,7 @@ GlobalOptions globalOptions(KompileOptions options) {
     OuterParsingOptions outerParsingOptions(KompileOptions options) { return options.outerParsing; }
+
+    @Provides
+    OutputDirectoryOptions outputDirectoryOptions(KompileOptions options) { return options.outputDirectory; }
```
`@RequestScoped`?

radumereuta commented on ocaml-backend/src/main/java/org/kframework/backend/ocaml/DefinitionToOcaml.java:
```diff
@@ -286,18 +286,18 @@ public void initialize(CompiledDefinition def) {
-            this.makeStuck = convert(new Kompile(kompileOptions, globalOptions, files, kem).compileRule(def.kompiledDefinition, makeStuck));
-            this.makeUnstuck = convert(new Kompile(kompileOptions, globalOptions, files, kem).compileRule(def.kompiledDefinition, makeUnstuck));
+            this.makeStuck = convert(new Kompile(kompileOptions, kompileOptions.outerParsing, globalOptions, files, kem).compileRule(def.kompiledDefinition, makeStuck));
+            this.makeUnstuck = convert(new Kompile(kompileOptions, kompileOptions.outerParsing, globalOptions, files, kem).compileRule(def.kompiledDefinition, makeUnstuck));
```
Probably these references should all use the injected outerParsing object instead of the reference from kompileOptions. That's going to be null.

dwightguth replied:
This doesn't actually matter because this method is always called by kompile, where it actually won't be null

radumereuta reviewed (approved):
Ok, I'm convinced by Dwight's answers.

rv-jenkins committed:
Merge branch 'master' into kprove-i-2
